// containers/Partners/index.js
import React from 'react';
import {connect} from 'react-redux';

import Header from 'components/Header';
import Footer from 'components/Footer';
import SimpleForm from 'components/Modal/SimpleForm';

import { sendModal } from 'actions';

import classNames from 'classnames';

let clearString = str => str.split('\n').join('').split('\t').join('').split('\r').join('');

class Partners extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      submit: false
    };

    this.ready = this.ready.bind(this);
    this.getDescription = this.getDescription.bind(this);
    this.getBlocks = this.getBlocks.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  ready(data) {
    return typeof data === 'object';
  }

  getDescription(str) {
    return clearString(str).match(/<p>(.*?)<\/p>/g).join('');
  }

  getBlocks(str) {
    return clearString(str).match(/<dl>(.*?)<\/dl>/g).map(block => {
      const dd = block.match(/<dd>(.*?)<\/dd>/)[1];
      return {
        title: block.match(/<dt>(.*?)<\/dt>/)[1],
        description: dd.match(/<span>(.*?)<\/span>/)[1],
        iconUrl: dd.match(/<img src="(.*?)" \/>/)[1]
      };
    });
  }

  handleSubmit(values) {
    const { siteSettings } = this.props;
    sendModal({ form: 'form-partner', ...values, 'csrf': siteSettings.token });
    this.setState({submit: true});
  }
  
  render() {
    const { submit } = this.state;
    const { page } = this.props;

    return (
      <React.Fragment>
        <Header/>
        <main className="main page-partners">
          <section className="page-partners__content ppar-content">
            <div className="container">
              <div className="row">
                <div className="col-12">
                  <h1 className="ppar-content__title">{ this.ready(page) && page.lang.ru.title }</h1>
                </div>
              </div>

              <div className="row">
                <div className="col-12">
                  {
                    this.ready(page) && <div
                      className="ppar-content__desc b-content"
                      dangerouslySetInnerHTML={{__html: this.getDescription(page.lang.ru.content)}}
                    />
                  }
                </div>
              </div>

              <div className="row">
                {
                  this.ready(page) && this.getBlocks(page.lang.ru.content).map((element, id) => {
                    const colClass = classNames(
                      'col-sm-6 col-lg-5',
                      {
                        'offset-lg-1': id % 2 === 1
                      }
                    );
                    return (
                      <div
                        key={id}
                        className={colClass}
                      >
                        <div className="ppar-content__item card-text-info">
                          <img src={element.iconUrl} className="card-text-info__icon" />
                          <h4 className="card-text-info__title">{element.title}</h4>
                          <div className="card-text-info__desc b-content">
                            <p>{element.description}</p>
                          </div>
                        </div>
                      </div>
                    );
                  })
                }
              </div>

              <div className="row">
                <div className="col-12">
                  <div className="ppar-content__form popup-wrap">
                    {
                      !submit && <React.Fragment>
                        <h3 className="popup-wrap__title">Хотите стать партнером?</h3>
                        <div
                          className="popup-wrap__desc b-content"
                          dangerouslySetInnerHTML={{__html: 'менеджер разъяснит все детали и подберет<br/>удобное время для визита к нам  офис'}}
                        />
                        <SimpleForm
                          onSubmit={this.handleSubmit}
                          button={"Отправить"}
                        />
                      </React.Fragment>
                    }
                    {
                      submit && <React.Fragment>
                        <h3 className="popup-wrap__title">Спасибо! Заявка отправлена.</h3>
                        <div
                          className="popup-wrap__desc b-content"
                          dangerouslySetInnerHTML={{__html: 'В ближайшее время наш менеджер свяжется с Вами!'}}
                        />
                      </React.Fragment>
                    }
                  </div>
                </div>
              </div>
            </div>
          </section>

        </main>
        <Footer />
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  siteSettings: state.siteSettings,
  page: state.pages['for-partners']
});

export default connect(mapStateToProps, null)(Partners);
