// components/MainFilter/filterField.js
import React, {Component} from 'react';
import classNames from 'classnames';

class FilterField extends Component {
  constructor(props) {
    super(props);
    this.state = {
      valueMin: props.valueMin,
      valueMax: props.valueMax,
      valueLengthMin: 0,
      valueLengthMax: 0
    };

    this.handlerMin = this.handlerMin.bind(this);
    this.handlerMax = this.handlerMax.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    const { valueMin, valueMax } = nextProps;
    this.setState({valueMin: valueMin});
    this.setState({valueMax: valueMax});
  }

  handlerMin(e) {
    const value = e.target.value !== '' ? e.target.value : '';
    this.props.onChangedMin(value);
    this.setState({valueMin: value, valueLengthMin: value.length});
  }

  handlerMax(e) {
    const value = e.target.value !== '' ? e.target.value : '';
    this.props.onChangedMax(value);
    this.setState({valueMax: value, valueLengthMax: value.length});
  }

  render() {
    const {valueMin, valueMax, valueLengthMin, valueLengthMax} = this.state;
    const {type, name, min, max} = this.props;

    const classListMin = classNames('form-field__label', { 'form-field__label_hide': valueLengthMin !== 0 });
    const classListMax = classNames('form-field__label', { 'form-field__label_hide': valueLengthMax !== 0 });

    return (
      <div className="card-filter-field">
        <p className="card-filter-field__title">{name}</p>
        <div className="card-filter-field__inputs">
          <div className="form-field form-field_left form-field_label card-filter-field__first-input">
            <label
              htmlFor={`${type}Min`}
              className={classListMin}
            >от</label>
            <input
              id={`${type}Min`}
              type="number"
              className="form-field__input"
              placeholder={min}
              min={min}
              max={max}
              value={valueMin}
              onChange={this.handlerMin}
            />
          </div>
          <div className="form-field form-field_right form-field_label card-filter-field__last-input">
            <label
              htmlFor={`${type}Max`}
              className={classListMax}
            >до</label>
            <input
              id={`${type}Max`}
              type="number"
              className="form-field__input"
              placeholder={max}
              min={min}
              max={max}
              value={valueMax}
              onChange={this.handlerMax}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default FilterField;
