// components/MainFilter/index.js
import React, {Component} from 'react';
import {connect} from 'react-redux';

import classNames from 'classnames';

import FilterField from 'components/MainFilter/filterField';
import MyMultiSelect from 'components/MyMultiSelect';
import DefaultItemRender from 'components/MyMultiSelect/defaultItemRender';
import DefaultValueSityRender from 'components/MyMultiSelect/defaultValueSityRender';
import DefaultValueConceptRender from 'components/MyMultiSelect/defaultValueConceptRender';

import {fetchConcept, fetchCity, selectFilter} from 'actions';
import {getConceptInFilter, getCityInFilter} from 'selectors';

class MainFilter extends Component {
	constructor() {
		super();
		this.state = {
			selectedSity: [],
			selectedConcept: [],
		};
  }
  
	componentDidMount() {
		this.props.fetchConcept();
		this.props.fetchCity();
	}

	handleSelectedSityChanged(selected) {
		this.setState({selectedSity: selected});
  }

	handleSelectedConceptChanged(selected) {
		this.setState({selectedConcept: selected});
  }
  
  render() {
    const {selectedSity, selectedConcept} = this.state;
    const {msCity, msConcept, pageMap, pageHome, pagePlace} = this.props;

    const filterClass = classNames({
      'home-filter': pageHome,
      'main-filter': pageMap || pagePlace,
      'page-home__filter': pageHome,
      'page-map__filter': pageMap,
      'page-site__filter': pagePlace
    });

    const filterContainerClass = classNames({
      'home-filter__container': pageHome,
      'main-filter__container': pageMap || pagePlace
    });

    const filterPriceClass = classNames({
      'home-filter__price': pageHome,
      'main-filter__price': pageMap || pagePlace
    });

    const filterAreaClass = classNames({
      'home-filter__area': pageHome,
      'main-filter__area': pageMap || pagePlace
    });

    const filterSityClass = classNames({
      'home-filter__sity': pageHome,
      'main-filter__sity': pageMap || pagePlace
    });

    const filterConceptClass = classNames({
      'home-filter__concept': pageHome,
      'main-filter__concept': pageMap || pagePlace
    });

    const filterButtonClass = classNames({
      'home-filter__button': pageHome,
      'main-filter__button': pageMap || pagePlace
    })

    return (
      <div className={filterClass}>
        <div className={filterContainerClass}>
          <div className={filterPriceClass}>
            <FilterField
              type="price"
              name="Цена, млн. руб"
              min="2.8"
              max="9.8"
            />
          </div>
          <div className={filterAreaClass}>
            <FilterField
              type="area"
              name="Площадь, сотки"
            />
          </div>
          <div className={filterSityClass}>
            <div className="card-filter-field">
              <p className="card-filter-field__title">Расположение</p>
              <div className="card-filter-field__inputs">
                <MyMultiSelect
                  addClass="card-filter-field__select"
                  options={msCity}
                  onSelectedChanged={this.handleSelectedSityChanged.bind(this)}
                  selected={selectedSity}
                  ItemRenderer={DefaultItemRender}
                  valueRenderer={DefaultValueSityRender}
                  disableSearch={true}
                  selectAllLabel="Все города"
                />
              </div>
            </div>
          </div>
          <div className={filterConceptClass}>
            <div className="card-filter-field">
              <p className="card-filter-field__title">Концепция</p>
              <div className="card-filter-field__inputs">
                <MyMultiSelect
                  addClass="card-filter-field__select"
                  options={msConcept}
                  onSelectedChanged={this.handleSelectedConceptChanged.bind(this)}
                  selected={selectedConcept}
                  ItemRenderer={DefaultItemRender}
                  valueRenderer={DefaultValueConceptRender}
                  disableSearch={true}
                  selectAllLabel="Все концепции"
                />
              </div>
            </div>
          </div>
          <div className={filterButtonClass}>
            <div className="card-filter-field">
              <p className="card-filter-field__title"></p>
              <div className="card-filter-field__inputs">
                <button className="btn btn_color_ac card-filter-field__button">Показать</button>
              </div>
            </div>
          </div>
        </div>
        <div className="main-filter__bg"></div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
	msConcept: getConceptInFilter(state),
  msCity: getCityInFilter(state)
});

const mapDispatchToProps = {
  fetchConcept,
  fetchCity
};

export default connect(mapStateToProps, mapDispatchToProps)(MainFilter);
