// components/MainFilter/index.js
import React, {Component} from 'react';

import classNames from 'classnames';

import FilterField from 'components/MainFilter/filterField';
import MyMultiSelect from 'components/MyMultiSelect';
import DefaultItemRender from 'components/MyMultiSelect/defaultItemRender';
import DefaultValueSityRender from 'components/MyMultiSelect/defaultValueSityRender';
import DefaultValueConceptRender from 'components/MyMultiSelect/defaultValueConceptRender';

const statesList = [
	{
		value: 'simpheropol',
		label: 'Симферополь'
	},
	{
		value: 'alushta',
		label: 'Алушта'
	},
	{
		value: 'yalta',
		label: 'Ялта'
	},
];

class MainFilter extends Component {
	constructor() {
		super();
		this.state = {
			selectedSity: [],
			selectedConcept: [],
		};
	}

	handleSelectedSityChanged(selected) {
		this.setState({selectedSity: selected});
  }

	handleSelectedConceptChanged(selected) {
		this.setState({selectedConcept: selected});
  }
  
  render() {
    const {selectedSity, selectedConcept} = this.state;
    const {pageMap, pageHome} = this.props;
    
    const filterClass = classNames({
      'home-filter': pageHome,
      'main-filter': pageMap,
      'page-home__filter': pageHome,
      'page-map__filter': pageMap
    });

    const filterContainerClass = classNames({
      'home-filter__container': pageHome,
      'main-filter__container': pageMap
    });

    const filterPriceClass = classNames({
      'home-filter__price': pageHome,
      'main-filter__price': pageMap
    });

    const filterAreaClass = classNames({
      'home-filter__area': pageHome,
      'main-filter__area': pageMap
    });

    const filterSityClass = classNames({
      'home-filter__sity': pageHome,
      'main-filter__sity': pageMap
    });

    const filterConceptClass = classNames({
      'home-filter__concept': pageHome,
      'main-filter__concept': pageMap
    });

    const filterButtonClass = classNames({
      'home-filter__button': pageHome,
      'main-filter__button': pageMap
    })

    return (
      <div className={filterClass}>
        <div className={filterContainerClass}>
          <div className={filterPriceClass}>
            <FilterField
              type="price"
              name="Цена, млн. руб"
              min="2.8"
              max="9.8"
            />
          </div>
          <div className={filterAreaClass}>
            <FilterField
              type="area"
              name="Площадь, сотки"
            />
          </div>
          <div className={filterSityClass}>
            <div className="card-filter-field">
              <p className="card-filter-field__title">Расположение</p>
              <div className="card-filter-field__inputs">
               <MyMultiSelect
                  addClass="card-filter-field__select"
                  options={statesList}
                  onSelectedChanged={this.handleSelectedSityChanged.bind(this)}
                  selected={selectedSity}
                  ItemRenderer={DefaultItemRender}
                  valueRenderer={DefaultValueSityRender}
                  disableSearch={true}
                  selectAllLabel="Все города"
                />
              </div>
            </div>
          </div>
          <div className={filterConceptClass}>
            <div className="card-filter-field">
              <p className="card-filter-field__title">Концепция</p>
              <div className="card-filter-field__inputs">
                <MyMultiSelect
                  addClass="card-filter-field__select"
                  options={statesList}
                  onSelectedChanged={this.handleSelectedConceptChanged.bind(this)}
                  selected={selectedConcept}
                  ItemRenderer={DefaultItemRender}
                  valueRenderer={DefaultValueConceptRender}
                  disableSearch={true}
                  selectAllLabel="Все концепции"
                />
              </div>
            </div>
          </div>
          <div className={filterButtonClass}>
            <div className="card-filter-field">
              <p className="card-filter-field__title"></p>
              <div className="card-filter-field__inputs">
                <button className="btn btn_color_ac card-filter-field__button">Показать</button>
              </div>
            </div>
          </div>
        </div>
        <div className="main-filter__bg"></div>
      </div>
    );
  }
}

export default MainFilter;
