// components/Modal/index.js
import React from 'react';
import ReactDOM from 'react-dom';
import {connect} from 'react-redux';

import classnames from 'classnames';
import { Close } from 'components/UiIcon';

import SimpleForm from './SimpleForm';

import { sendModal } from 'actions';

class Modal extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      submit: false
    };

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentWillMount() {
    this.root = document.createElement('div');
    this.root.className = 'portal'
    document.body.appendChild(this.root);
  }

  componentWillUnmount() {
    document.body.removeChild(this.root);
  }

  handleSubmit(values) {
    const { onClose, siteSettings } = this.props;
    sendModal({ ...this.props.data, ...values, 'csrf': siteSettings.token });
    this.setState({submit: true});
    setTimeout(() => {
      onClose();
    }, 3000);
  }

  render() {
    const { submit } = this.state;
    const { title, desc, button, display, visible, onClose } = this.props;

    const popupClassName = classnames(
      'popup', {
        'display': display,
        'visible': visible
      }
    )

    const wrapClassName = classnames(
      'popup__wrap',
      'popup-wrap',
      {
        'popup-wrap_submit': submit
      }
    )

    return ReactDOM.createPortal(
      <div className={popupClassName}>
        <div
          className="popup__bg"
          onClick={onClose}
        />
        <div className={wrapClassName}>
          <button
            className="popup-wrap__close"
            onClick={onClose}
          >
            <Close />
          </button>
          {
            !submit && <React.Fragment>
              <h3 className="popup-wrap__title">{title}</h3>
              <div
                className="popup-wrap__desc b-content"
                dangerouslySetInnerHTML={{__html: desc}}
              />
              <SimpleForm
                onSubmit={this.handleSubmit}
                button={button}
              />
            </React.Fragment>
          }
          {
            submit && <React.Fragment>
              <h3 className="popup-wrap__title">Спасибо! Заявка отправлена.</h3>
              <div
                className="popup-wrap__desc b-content"
                dangerouslySetInnerHTML={{__html: desc}}
              />
            </React.Fragment>
          }
        </div>
      </div>,
      this.root
    );
  }
}

const mapStateToProps = state => ({
  siteSettings: state.siteSettings
});

export default connect(mapStateToProps, null)(Modal);
