// components/MyMultiSelect/index.js
import React, {Component} from 'react';

import classNames from 'classnames';

import Dropdown from './dropdown.js';
import SelectPanel from './select-panel.js';

class MyMultiSelect extends Component {
	static defaultProps = {
		hasSelectAll: true,
		shouldToggleOnHover: false,
	}

	getSelectedText() {
		const {options, selected} = this.props;

		const selectedOptions = selected
			.map(s => options.find(o => o.value === s));

		const selectedLabels = selectedOptions.map(s => s ? s.label : "");

		return selectedLabels.join(", ");
	}

	renderHeader() {
		const {
			options,
			selected,
			valueRenderer,
		} = this.props;

		const noneSelected = selected.length === 0;
		const allSelected = selected.length === options.length;

		const customText = valueRenderer && valueRenderer(selected, options);

		if (noneSelected) {
			return <span style={styles.noneSelected}>
				{customText || "Select some items..."}
			</span>;
		}

		if (customText) {
			return <span>{customText}</span>;
		}

		return <span>
			{allSelected
				? "All items are selected"
				: this.getSelectedText()
			}
		</span>;
	}

	handleSelectedChanged = (selected) => {
		const {onSelectedChanged, disabled} = this.props;

		if (disabled) {
			return;
		}

		if (onSelectedChanged) {
			onSelectedChanged(selected);
		}
	}

	render() {
		const {
			addClass,
			ItemRenderer,
			options,
			selected,
			selectAllLabel,
			isLoading,
			disabled,
			disableSearch,
			filterOptions,
			shouldToggleOnHover,
			hasSelectAll,
		} = this.props;

		return <div className={classNames('multi-select', addClass)}>
			<Dropdown
				isLoading={isLoading}
				contentComponent={SelectPanel}
				shouldToggleOnHover={shouldToggleOnHover}
				contentProps={{
					ItemRenderer,
					options,
					selected,
					hasSelectAll,
					selectAllLabel,
					onSelectedChanged: this.handleSelectedChanged,
					disabled,
					disableSearch,
					filterOptions,
				}}
				disabled={disabled}
			>
				{this.renderHeader()}
			</Dropdown>
		</div>;
	}
}

const styles = {
	noneSelected: {
		color: "#aaa",
	},
};

export default MyMultiSelect;
export {Dropdown};
