// containers/Contacts/Map.js
import React, { Component } from 'react';
import classnames from 'classnames';

import pinIcon from 'components/UiIcon/pin.svg';

let myMap, points;
const yMaps = window.ymaps;

class Map extends Component {
  constructor(props) {
    super(props);
  }
  
  componentWillMount() {
    yMaps.ready(() => {
      myMap = new yMaps.Map('mapcontacts', {
          center: [44.862862, 34.369187],
          zoom: 9,
          controls: ['zoomControl', 'fullscreenControl']
        }, {
          searchControlProvider: 'yandex#search'
        });

      myMap.setCenter([44.862862, 34.369187], 9); 
    });
  }

  componentWillUpdate(nextProps) {
    yMaps.ready(() => {
      myMap = new yMaps.Map('mapcontacts', {
          center: [44.862862, 34.369187],
          zoom: 9,
          controls: ['zoomControl', 'fullscreenControl']
        }, {
          searchControlProvider: 'yandex#search'
        });

      myMap.setCenter([44.862862, 34.369187], 9); 
    });
  }

  render() {
    const { className } = this.props;

    const mapYMapClass = classnames(
      'card-contacts-map__body',
      className
    );

    return (
      <div
        className={mapYMapClass}
        id="mapcontacts"
      ></div>
    );
  }
}

export default Map;
