// containers/Array/ArrayPlots.js
import React from 'react';
import Modal from 'components/Modal';

import { CardSite } from 'components/CardSite';

class ArrayPlots extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      displayModal: false,
      visibleModal: false,
      dataModal: {}
    };

    this.togglePopup = this.togglePopup.bind(this);
  }

  togglePopup(id) {
    const { displayModal, visibleModal } = this.state;

    if(!displayModal) {
      this.setState({
        displayModal: !displayModal,
        dataModal: {
          form: 'form-land-plots',
          land_plots_id: id
        }
      });
      setTimeout(() => {
        this.setState({visibleModal: !visibleModal});
      }, 5);
    } else {
      this.setState({visibleModal: !visibleModal});
      setTimeout(() => {
        this.setState({displayModal: !displayModal});
      }, 200);
    }
  }

  render() {
    const { arrayPlots } = this.props;
    const { displayModal, visibleModal, dataModal } = this.state;
    return (
      <React.Fragment>
        <section className="page-array__site parr-site">
          <div className="container">
            <div className="row">
              <div className="col-12">
                {
                  arrayPlots.length !== 0 && arrayPlots.map((element, id) => {
                    return <CardSite
                      key={id}
                      id={element.id}
                      name={element.name}
                      adress={element.adress}
                      area={element.area}
                      price={element.price_total}
                      openPopup={this.togglePopup}
                    />
                  })
                }
                
              </div>
            </div>
          </div>
        </section>
        {
          displayModal && <Modal
            display={displayModal}
            visible={visibleModal}
            data={dataModal}
            title={"Чтобы забронировать участок, оставьте заявку"}
            desc={"<p>менеджер разъяснит все детали и подберет<br /> удобное время для визита к нам в офис</p>"}
            onClose={this.togglePopup}
          />
        }
      </React.Fragment>
    );
  }

}

export default ArrayPlots;
