// containers/Array/ArrayMain.js
import React from 'react';
import Modal from 'components/Modal';

class ArrayMain extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      displayModal: false,
      visibleModal: false,
      dataModal: {}
    };

    this.togglePopup = this.togglePopup.bind(this);
  }

  togglePopup(id) {
    const { displayModal, visibleModal } = this.state;

    if(!displayModal) {
      this.setState({
        displayModal: !displayModal,
        dataModal: {
          form: 'form-array',
          array_id: id
        }
      });
      setTimeout(() => {
        this.setState({visibleModal: !visibleModal});
      }, 5);
    } else {
      this.setState({visibleModal: !visibleModal});
      setTimeout(() => {
        this.setState({displayModal: !displayModal});
      }, 200);
    }
  }

  scrollToPlots() {
    const V = 0.3;
    const w = window.pageYOffset;
    const t = document.querySelector('.parr-site').getBoundingClientRect().top;
    let start = null;
    requestAnimationFrame(step);  // подробнее про функцию анимации [developer.mozilla.org]
    function step(time) {
      if (start === null) start = time;
      let progress = time - start,
        r = (t < 0 ? Math.max(w - progress/V, w + t) : Math.min(w + progress/V, w + t));
      window.scrollTo(0,r);
      if (r != w + t) {
        requestAnimationFrame(step)
      }
    }
  }

  render() {
    const {id, name, mainImg} = this.props;
    const { displayModal, visibleModal, dataModal } = this.state;

    return (
      <React.Fragment>
        <section className="page-array__main parr-main">
          <img src={mainImg} className="parr-main__bg" />
          
          <div className="container parr-main__container">
            <div className="row">
              <div className="col-lg-8">
                <h1 className="parr-main__title">{name}</h1>
              </div>
              <div className="col-lg-4">
                <button
                  className="btn btn_color_ac parr-main__btn-buy"
                  onClick={() => this.togglePopup(id)}
                >Купить массив</button>
                <button
                  className="btn btn_color_white parr-main__btn-view-site"
                  onClick={this.scrollToPlots}
                >Смотреть участки</button>
              </div>
            </div>
          </div>
        </section>
        {
          displayModal && <Modal
            display={displayModal}
            visible={visibleModal}
            data={dataModal}
            title={"Чтобы забронировать массив, оставьте заявку"}
            desc={"<p>менеджер разъяснит все детали и подберет<br /> удобное время для визита к нам в офис</p>"}
            button={"Забронировать"}
            onClose={this.togglePopup}
          />
        }
      </React.Fragment>
    );
  }
}

export default ArrayMain;
